File: arduino/communication/Protocol.cpp
#include "Arduino.h"
#include "Protocol.h"

void Protocol::Send(String s)
{
  Serial.println(s);
}

void Protocol::Read()
{
  String readString = Serial.readStringUntil(10);
  this->readString = readString;
}

void Protocol::SendAlive()
{
  if (millis() - this->lastAliveMessage >= this->diffBetweenAliveMessages || (this->isAlive == false && millis() - this->lastAliveMessage >= this->diffBetweenAliveMessages / 100))
  {
    if (this->hasSentAliveYet == false)
    {
      this->isAlive = false;
    }
    this->Send("CHECK-ALIVE");
    this->lastAliveMessage = millis();
    this->hasSentAliveYet = false;
  }
}

void Protocol::HandleRead()
{
  if (Serial.available() > 0)
  {
    this->Read();
    if (this->isAlive == true)
    {
      if (this->readString.equals("turn led on"))
      {
        this->TurnLedOn();
      }
      else if (this->readString.equals("turn led off"))
      {
        this->TurnLedOff();
      }
      else if (this->hasSentAliveYet = false && this->readString.equals("VERIFY-ALIVE"))
      {
        this->hasSentAliveYet = true;
        this->Send("VERIFIED-COMMUNICATION");
      }
    }
    else
    {
      if (this->readString.equals("VERIFY-ALIVE"))
      {
        this->isAlive = true;
        this->hasSentAliveYet = true;
        this->Send("VERIFIED-COMMUNICATION");
      }
    }
  }
}

void Protocol::TurnLedOn()
{
  this->ledOn = 1;
}

void Protocol::TurnLedOff()
{
  this->ledOn = 0;
}

void Protocol::SendButtonPressed()
{
  this->Send("button was pressed");
}

File: pi/Protocol.py
import io


class Protocol:
    isAlive = 0
    ser = None
    lastRead = None
    lastReadString = ""
    buttonPressedTimes = 0
    led = 0

    def setSerial(self, ser):
        self.ser = ser

    def sendAlive(self):
        self.send("VERIFY-ALIVE")

    def send(self, value):
        self.ser.write((str(value)).encode())

    def read(self):
        try:
            self.lastRead = self.ser.readline()
            self.lastReadString = str(
                self.lastRead, 'utf-8').replace('\r', '').replace("\n", "")
            return self.lastReadString
        except:
            try:
                print("couldnt transform into utf8: "+str(self.lastRead))
            except:
                print("couldnt handle string")
            return ""

    def handleRead(self):
        readString = self.read()
        print(readString)
        if readString == str("CHECK-ALIVE"):
            self.sendAlive()
            self.isAlive = 1
        elif readString == str("button was pressed"):
            self.buttonPressedTimes = self.buttonPressedTimes + 1
        elif readString == str("led is on"):
            self.led = 1
        elif readString == str("led is off"):
            self.led = 0

    def getButtonPressedTimes(self):
        return self.buttonPressedTimes


protocolInstance = Protocol()
